walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,96 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // Using linked list as removing the first element will take O(1) time
+    LinkedList<Event> eventList = new LinkedList<Event>(events);
+    // Collections.sort() -> O(nlogn)
+    Collections.sort(eventList, Event.ORDER_EVENT_BY_START);
+
+    ArrayList<TimeRange> openRanges = new ArrayList<TimeRange>();
+
+    // making sure the request is within bounds
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return openRanges;
+    }
+
+    if (request.getDuration() < 0) {
+      return openRanges;
+    }
+    return traverseThroughSchedule(events, request, eventList, openRanges);
+  }
+
+  public Collection<TimeRange> traverseThroughSchedule(
+      Collection<Event> events,
+      MeetingRequest request,
+      LinkedList<Event> eventList,
+      ArrayList<TimeRange> openRanges) {
+
+    long currentTime = (long) TimeRange.START_OF_DAY;
+
+    while (currentTime < (long) TimeRange.END_OF_DAY) {
+
+      // if the last open time slot is between the last meeting and the end of the day
+      if (eventList.isEmpty()) {
+        if ((long) TimeRange.END_OF_DAY - currentTime < request.getDuration()) {
+          break;
+        }
+        addFreeSpace(currentTime, TimeRange.END_OF_DAY, openRanges, true);
+        break;
+      }
+
+      // Get the first element
+      Event topOfList = eventList.getFirst();
+
+      // if there are no common people between the request and the event
+      if (Collections.disjoint(topOfList.getAttendees(), request.getAttendees())) {
+        eventList.removeFirst();
+        continue;
+      }
+
+      if ((currentTime + request.getDuration()) <= topOfList.getStartTime()) {
+        addFreeSpace(currentTime, (int) topOfList.getStartTime(), openRanges, false);
+        currentTime = topOfList.getStartTime();
+        continue;
+      } else {
+        if (Collections.disjoint(topOfList.getAttendees(), request.getAttendees())) {
+          checkIfSpaceInMeeting(topOfList, currentTime, request, openRanges, eventList);
+          continue;
+        } else {
+          long removedTime = eventList.removeFirst().getEndTime();
+          currentTime = currentTime < removedTime ? removedTime : currentTime;
```

Comment:
Again, this should be an if statement instead of a ternary operator. Also add a comment explaining why you're doing this. It seems to me that since you sort by start time, the end of the last event may be after the end of this current event and you'll always want currentTime to be the max of those values.

Reply:
Yes exactly